Renewable internal structures, TS loading [1/3]

flomnes opened the pull request.

flomnes committed:
Rework simulation.cpp's mapping

flomnes committed:
Add renewable data structures

flomnes committed:
Defines / prepro

flomnes committed:
Enum -> string conversion function

flomnes committed:
Remove coupling

flomnes committed:
Fix cluster.cpp

flomnes committed:
Fix build

flomnes committed:
Remove prepro files

flomnes committed:
Format sources

flomnes committed:
Load renewable TS

This reverts commit 4ae02b22f83812d027353861cc2f0cabeb761950.

flomnes committed:
Merge remote-tracking branch 'origin/develop' into feature/renewable_internal_structures

flomnes committed:
Fix build

flomnes committed:
Add roof-top and "other" 2, 3 and 4

flomnes committed:
Remove debug message

flomnes committed:
Look for renewable data in input/renewables (not renewable)

flomnes committed:
Fix warnings related to renewable's area.ini

flomnes committed:
Fix container.hxx

flomnes committed:
Residual load: include renewable clusters

flomnes committed:
Fix TS alloc

flomnes committed:
Area: call renewable actions

flomnes committed:
Remove obsolete comments

flomnes committed:
Flush renewable TS (swap flavour)

flomnes committed:
Add TODO note

flomnes committed:
Remove unused code

Renewable clusters did not exist in versions < 390

flomnes committed:
Force loading of renewable TS

Since TS generation is not possible

guilpier-code commented on src/libs/antares/study/area/area.hxx:
```diff
@@ -61,7 +61,9 @@ inline bool Area::storeTimeseriesNumbers(Study& study)
         return storeTimeseriesNumbersForThermal(study);
-
+        // Renewable
+    case timeSeriesRenewable:
```
It is possible that current function (**Area::storeTimeseriesNumbers(Study& study)**) is never called for renewables.

flomnes replied:
Fixed in commit 7b11174

flomnes replied:
It is called after PR #251, added in commit 4be8e9f41.

guilpier-code commented on src/libs/antares/study/area/list.cpp:
```diff
@@ -1212,6 +1232,28 @@ bool AreaList::loadFromFolder(const StudyLoadOptions& options)
+    // Renewable data, specific to areas
+    {
+        logs.info() << "Loading renewable clusters...";
+        buffer.clear() << pStudy.folderInput << SEP << "renewables" << SEP << "areas.ini";
```
For thermal clusters, this areas.ini file contains unsupplied and spilled energy costs.
I don't know why this information is located inside thermal directory, as it does not regard thermal only. Maybe there is reason, but I do not see it now.
Anyway, for renewable, this information does not seem to have any meaning.

flomnes replied:
Right.

This is only an hypothesis, but energy inadequacies (spilled or unsupplied) are used to determine which thermal cluster to use.

However, the hydro part is also controllable.

Therefore, spillage & unsupplied costs should be placed in a per-area directory.

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+                    delete cluster;
+                    continue;
+                }
+                // keeping track of the cluster
```
For now, some load of modulations (marginal cost, capacity, ...) are missing. Will be added later.

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+                    continue;
+                }
+                // keeping track of the cluster
+                mapping[cluster->id()] = cluster;
```
Why do we keep 2 maps with clusters (**mapping** and the one we populate in **add(cluster)** instruction).
Maybe useful for thermal clusters (eventually to separate must-run clusters from others in my memory), not sure it is for renewable clusters.

flomnes replied:
`mapping` removed in commit d9895347

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+                // keeping track of the cluster
+                mapping[cluster->id()] = cluster;
+
+                cluster->flush();
```
As flush function does nothing, why do we flush ?

flomnes replied:
In the swap flavor, some actions are required for time-series.

guilpier-code replied:
ok, but **cluster->flush()** still does nothing. According to me, this instruction should be removed and the definition : 
```
#ifndef ANTARES_SWAP_SUPPORT
inline void RenewableCluster::flush()
{
    // do nothing
}
#endif
```
should be removed as well, for renewable and as well as its thermal clusters counterpart.

flomnes replied:
Calling `flush` on a cluster (whatever the type) calls `flush` on the `series` and `timeseriesNumbers` objects. See **src/libs/antares/study/parts/common/series.hxx** :

```cpp
inline void DataSeriesCommon::flush()
{
#ifdef ANTARES_SWAP_SUPPORT
    series.flush();
    timeseriesNumbers.flush();
#endif
}
```

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+            if (not cluster.group().empty())
+                s->add("group", cluster.group());
+            if (not cluster.enabled)
+                s->add("enabled", "false");
```
For now, some parameters are missing (nominal capacity, marginal cost, ...), but they will be taken into account when specification are more advanced.

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+    // timseries
+    series->series = cluster.series->series;
+    cluster.series->series.unloadFromMemory();
+    series->timeseriesNumbers.clear();
```
ok, probably some missing parameters

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+    mapping.clear();
+
+    if (not cluster.empty())
+        cluster.clear();
```
in the case of renewable clusters, **mapping** and **cluster** lists are probably redundants.
They are not destroyed in the same way but it's relevant, as values of these maps are addresses to clusters.

flomnes replied:
`mapping` is redundant indeed. Fixed in commit d9895347.

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+        cluster.clear();
+}
+
+const RenewableCluster* RenewableClusterList::find(const RenewableCluster* p) const
+{
+    auto end = cluster.end();
+    for (auto i = cluster.begin(); i != end; ++i)
+    {
+        if (p == i->second)
+            return i->second;
+    }
+    return nullptr;
+}
+
+Data::RenewableCluster* RenewableClusterList::find(const RenewableCluster* p)
+{
+    auto end = cluster.end();
+    for (auto i = cluster.begin(); i != end; ++i)
+    {
+        if (p == i->second)
+            return i->second;
+    }
+    return nullptr;
+}
```
one of the 2 **find** methods is useless : if we comment the non const method, Antares still compiles.
Moreover, if **cluster.h** is touched, the recompilation is very long : a lot of files are recompiled. Why is that ? (maybe it's expected, to be checked).
Both remarks apply to thermal clusters as well, of course.

flomnes replied:
**part.h** is included in many places, it includes **cluster.h** indirectly. Thus, whenever it is changed, the recompilation takes time.

One method is `const`, the other is not. The non-`const` method is probably not used.

guilpier-code replied:
why do we not remove it then ?

flomnes replied:
In case we might need it later ?

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+    pGroup = newgrp;
+    newgrp.toLower();
+
+    switch (newgrp[0])
```
Cannot figure out why we start testing **newgrp[0]** before finally testing **newgrp**. Why not testing **newgrp** directly ? It would be more simple.

flomnes replied:
This is a clunky optimization. Removed in commit f6c8b85

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+    return false;
+}
+
+int RenewableClusterListSaveDataSeriesToFolder(const RenewableClusterList* l,
```
Not called yet

flomnes replied:
Right, fixed in commit 394a48b.

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+    return ret;
+}
+
+int RenewableClusterListSaveDataSeriesToFolder(const RenewableClusterList* l,
```
It seems this function won't be called, as it is called for thermal clusters when storing generated TS. So it should be removed.

flomnes replied:
I disagree. For thermal clusters, there are 2 calls :
> libs/antares/study/study.importprepro.cpp:122:23:                ret = ThermalClusterListSaveDataSeriesToFolder(
> libs/antares/study/area/list.cpp:335:15:        ret = ThermalClusterListSaveDataSeriesToFolder(&area.thermal.list, buffer) and ret;

The first call is used to export generated TS into the input folder. It does not make sense for renewables, as long as no TS generation is possible.

The second call is used e.g when the user changes TS values in the GUI (function `AreaList::saveToFolder`, called by `Study::saveToFolder`). This call makes sense for renewable clusters.

See commit 394a48b.

guilpier-code replied:
ok, my mistake

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,837 @@
+    if (parentArea)
+        parentArea->invalidate();
+}
+
```
Cluster and list of clusters are defined in the same files cluster.h/hxx/cpp. A separation would be clearer.

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.hxx:
```diff
@@ -0,0 +1,131 @@
+*/
+#ifndef __ANTARES_LIBS_STUDY_PARTS_RENEWABLE_CLUSTER_HXX__
+#define __ANTARES_LIBS_STUDY_PARTS_RENEWABLE_CLUSTER_HXX__
+
```
As already said, we should split file cluster.* into cluster.* and clusterList.*

guilpier-code commented on src/libs/antares/study/parts/renewable/container.h:
```diff
@@ -0,0 +1,120 @@
+    **
+    ** \return The number of disabled clusters found
+    */
+    uint removeDisabledClusters();
```
For now, this function is not used

flomnes replied:
Fixed in commit 7b11174.

guilpier-code commented on src/libs/antares/study/parts/renewable/container.cpp:
```diff
@@ -0,0 +1,140 @@
+    }
+}
+
+uint PartRenewable::removeDisabledClusters()
```
why is this function so complicated, with 2 loops (**do** and **for**) intertwined ?

flomnes replied:
These two loops are used to remove every disabled cluster from the list.

Iterator `list.end()` can change and become invalid if `list.remove(i);` is called. In this case, we get out of the `for` (with the `break` instruction) loop to reset this iterator.

Since `list.remove(i)` invalidates `list.end()`, the two loops are needed. See **src/libs/antares/study/parts/renewable/cluster.h**

```cpp
/*!
** \brief Remove a cluster represented by an iterator
**
** The renewable cluster will be removed from the list but _not_
** destroyed.
** The iterator should considered as invalid after using this method.
** \return void
*/
void remove(iterator i);
```

flomnes replied:
Re-written in pair programming.

guilpier-code commented on src/libs/antares/study/parts/renewable/series.h:
```diff
@@ -0,0 +1,107 @@
+/*!
+** \brief Data series (Renewable)
+*/
+class DataSeriesRenewable
```
This class does not seem to be different from its thermal counterpart.

flomnes replied:
Fixed in commits 949607a and bfb874d.

guilpier-code commented on src/solver/aleatoire/alea_tirage_au_sort_chroniques.cpp:
```diff
@@ -93,6 +93,27 @@ static void InitializeTimeSeriesNumbers_And_ThermalClusterProductionCost(
         }
+        // Renewable
+        {
+            auto end = area.renewable.list.mapping.end();
```
to be checked if **mapping** and **cluster** list are redundant

flomnes replied:
`mapping` is redundant. Removed in commit d9895347

guilpier-code commented on src/ui/simulator/toolbox/components/datagrid/renderer/simulation.cpp:
```diff
@@ -33,6 +33,16 @@
+// Anonymous namespace: global variable, local scope
+namespace
+{
+enum Antares::Data::TimeSeries mapping[] = {Data::timeSeriesLoad,
```
This file is part of GUI, but ok. Why do we not add **Data::timeSeriesRenewable** then ?

guilpier-code commented on src/libs/antares/study/fwd.h:
```diff
@@ -68,6 +68,8 @@ typedef Yuni::CString<ant_k_area_name_max_length * 2 + 1, false> AreaLinkName;
 //! Name of a single thermal
 typedef Yuni::CString<ant_k_cluster_name_max_length, false> ThermalClusterName;
 
+typedef Yuni::CString<ant_k_cluster_name_max_length, false> RenewableClusterName;
```
**ClusterName** should be enough for both thermal and renewable

flomnes committed:
{Thermal,Renewable}ClusterName -> ClusterName

flomnes committed:
UI : Add timeSeriesRenewable to mapping

flomnes committed:
Remove disabled renewable clusters before solver is called

flomnes committed:
Remove log space

flomnes committed:
Remove RenewableClusterList::mapping

Redundant with RenewableClusterList::cluster

flomnes committed:
Remove switch...case in string -> groupID conversion

Clunky optimization

flomnes committed:
Write renewable cluster list and TS

flomnes committed:
Use a common data structure for renewable & thermal TS

flomnes committed:
Apply clang-format

guilpier-code commented on src/libs/antares/study/study.cpp:
```diff
@@ -1122,8 +1129,8 @@ bool Study::thermalClusterRename(ThermalCluster* cluster, ThermalClusterName new
     // Otherwise the values associated to the area will be lost.
     logs.info() << "  renaming thermal cluster '" << cluster->name() << "' into '" << newName
                 << "'";
-    //	ThermalClusterName oldId   = cluster->id();
```
Changes in commented lines, should not we simply remove these lines ?

flomnes committed:
Remove DataSeries{Thermal,Renewable}

flomnes committed:
Improve memory estimation for renewable TS

guilpier-code commented on src/libs/antares/study/parts/common/series.h:
```diff
@@ -0,0 +1,73 @@
+/*!
+** \brief Data series (Common)
+*/
+class DataSeriesCommon
```
I would call this class DataSeriesCommonForClusters for instance, otherwise it could be about any TS

flomnes replied:
It may be extended to other modes as well (wind, solar, etc.)

guilpier-code commented on src/libs/antares/study/parts/common/series.cpp:
```diff
@@ -0,0 +1,75 @@
+{
+namespace Data
+{
+bool DataSeriesCommon::invalidate(bool reload) const
```
Same remark as in the related *.h

guilpier-code commented on src/libs/antares/study/parts/common/series.hxx:
```diff
@@ -24,27 +24,27 @@
 **
 ** SPDX-License-Identifier: licenceRef-GPL3_WITH_RTE-Exceptions
 */
-#ifndef __ANTARES_LIBS_STUDY_PARTS_THERMAL_TIMESERIES_HXX__
```
Same remark as in the related *.h, except that, here, **__ANTARES_LIBS_STUDY_PARTS_COMMON_TIMESERIES_HXX__** should be named otherwise accordingly

guilpier-code commented on src/libs/antares/study/parts/common/series.cpp:
```diff
@@ -0,0 +1,75 @@
+    timeseriesNumbers.estimateMemoryUsage(u, true, 1, u.years);
+    uint nbTimeSeries;
+    switch (ts) {
+    case timeSeriesThermal:
```
As **estimateMemoryUsage** method differs depending on the kind of cluster we're on, we perhaps should make 2 data series classes (renewable and thermal), which both take a DataSeriesCommonCluster as ancester, but differ with this method only. If so, we woud not have the second argument and the switch in the body.
Further more, this file could contain specialized functions **DataSeriesSaveToFolder** and **DataSeriesLoadFromFolder**.
Is it worth it ?

flomnes replied:
Probably not

guilpier-code commented on src/libs/antares/study/parts/renewable/series.cpp:
```diff
@@ -0,0 +1,90 @@
+{
+namespace Data
+{
+int DataSeriesSaveToFolder(const DataSeriesCommon* t,
```
See remark from **src/libs/antares/study/parts/common/series.cpp**

flomnes replied:
May be used for solar, load, etc. in future revisions.

guilpier-code commented on src/libs/antares/study/parts/renewable/series.h:
```diff
@@ -0,0 +1,69 @@
+** \param folder The target folder
+** \return A non-zero value if the operation succeeded, 0 otherwise
+*/
+int DataSeriesLoadFromFolder(Study& s,
```
See remark from **src/libs/antares/study/parts/common/series.cpp**

guilpier-code commented on src/libs/antares/study/parts/thermal/series.cpp:
```diff
@@ -40,9 +40,9 @@ namespace Antares
-int DataSeriesThermalSaveToFolder(const DataSeriesThermal* t,
-                                  const ThermalCluster* ag,
-                                  const AnyString& folder)
+int DataSeriesSaveToFolder(const DataSeriesCommon* t,
```
Same remark as above

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,805 @@
+    delete series;
+}
+
+#ifdef ANTARES_SWAP_SUPPORT
```
flush should be removed as well as its call

flomnes replied:
flush triggers some actions on the TS in the swap flavor. I insist it should be kept.

flomnes committed:
Split sources for RenewableCluster and RenewableClusterList

flomnes committed:
Add missing cluster_list.hxx

flomnes committed:
Re-implement removeDisabledClusters()

flomnes committed:
Refactor class Cluster (1/2)

* Create common class Cluster
* Derive ThermalCluster and RenewableCluster from it
* Rename `group` setter method to `setGroup` to avoid conflicting names
* Rename `name` setter method to `setName` (same reason)

TODO : move load/save free functions into classes as member functions.

flomnes committed:
Apply clang-format

flomnes committed:
Move load/save methods into RenewableClusterList

flomnes committed:
Move DataSeries load/save functions into Cluster class

and use in the RenewableClusterList class

flomnes committed:
Use Cluster load/save methods for thermal clusters

flomnes committed:
Virtual function Cluster::flush()

flomnes committed:
Fix simple-study

Only breaks Windows CI...

flomnes committed:
Make common ClusterList, use for renewables

flomnes committed:
Use std::vector<uint> instead of uint* for groupCount

flomnes committed:
Make ClusterList more generic

* Move container-related functions such as `find`, `remove` to template.
* Specialize functions only when needed.

guilpier-code commented on src/libs/antares/study/parts/common/cluster.h:
```diff
@@ -0,0 +1,143 @@
+public:
+    Cluster(Area* parent);
+
+    //! Get the thermal cluster ID
```
some "thermal" (inside comments) left in this file

guilpier-code commented on src/libs/antares/study/parts/thermal/cluster.h:
```diff
@@ -122,12 +120,12 @@ class ThermalCluster final : public Yuni::NonCopyable<ThermalCluster>
     /*!
     ** \brief Invalidate all data associated to the thermal cluster
     */
-    bool invalidate(bool reload) const;
```
Few lines above, function **FlexibilityIsValid** is never used. to be removed

guilpier-code commented on src/libs/antares/study/parts/thermal/cluster.h:
```diff
@@ -122,12 +120,12 @@ class ThermalCluster final : public Yuni::NonCopyable<ThermalCluster>
-    void markAsModified() const;
+    void markAsModified() const override;
 
     /*!
```
few lines underneath, function **invalidateArea** should be moved to common cluster class

guilpier-code commented on src/libs/antares/study/parts/thermal/cluster.h:
```diff
@@ -247,18 +231,6 @@ class ThermalCluster final : public Yuni::NonCopyable<ThermalCluster>
     */
```
Above, function **isVisibleOnLayer** should be declared and defined in the cluster common part, isn't it ?

flomnes replied:
done

guilpier-code commented on src/libs/antares/study/parts/thermal/cluster.h:
```diff
@@ -430,369 +387,9 @@ class ThermalCluster final : public Yuni::NonCopyable<ThermalCluster>
-                                             const AnyString& folder,
-                                             const YString& msg);
-
 struct CompareThermalClusterName final
```
**CompareThermalClusterName** should be re-defined as **CompareClusterName** and moved to common cluster part

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,267 @@
+
+bool Data::RenewableCluster::invalidate(bool reload) const
+{
+    bool ret = true;
```
why not invalidate modulation as well (like in thermal clusters) ?
`ret = modulation.invalidate(reload) and ret;`

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,267 @@
+void Data::RenewableCluster::markAsModified() const
+{
+    if (series)
+        series->markAsModified();
```
`modulation.markAsModified();` missing ?

guilpier-code commented on src/libs/antares/study/parts/common/cluster.h:
```diff
@@ -0,0 +1,143 @@
+    //! \warning this variable is only valid when used by the solver
+    // (initialized in the same time that the runtime data)
+    uint areaWideIndex;
+
```
**marginalCost** should be a common data, not only thermal cluster's

guilpier-code commented on src/libs/antares/study/parts/renewable/cluster.cpp:
```diff
@@ -0,0 +1,267 @@
+}
+
+void Data::RenewableCluster::reset()
+{
```
there should be a common part for reset, that is a **Cluster::reset()** and it should be called in both thermal and renewable cluster **reset()**

guilpier-code commented on src/libs/antares/study/study.cpp:
```diff
@@ -1071,7 +1078,7 @@ bool Study::areasThermalClustersMinStablePowerValidity(
 }
 
-bool Study::thermalClusterRename(ThermalCluster* cluster, ThermalClusterName newName, bool)
+bool Study::thermalClusterRename(Cluster* cluster, ClusterName newName, bool)
```
- so the function should be renamed as **clusterRename**
- Why is this function part of the study class ? Shouldn't it be a method of common class for clusters ? Seems a bit complicated to move this method. To be checked.

flomnes replied:
For now, this function should be duplicated. It would be somewhat tricky to de-duplicate.

guilpier-code commented on src/libs/antares/study/parts/common/cluster.h:
```diff
@@ -0,0 +1,143 @@
+class Cluster
+{
+public:
+    //! Set of renewable clusters
+    typedef std::set<Cluster*, CompareClusterName> Set;
+    //! Set of renewable clusters (pointer)
+    typedef std::set<Cluster*> SetPointer;
+    //! Map of renewable clusters
+    typedef std::map<ClusterName, Cluster*> Map;
+    //! Vector of renewable clusters
+    typedef std::vector<Data::Cluster*> Vector;
```
All these **typdef**s are useless, because unused.

guilpier-code commented on src/libs/antares/CMakeLists.txt:
```diff
@@ -195,22 +195,43 @@ set(SRC_STUDY_AREA_UI
+	study/parts/common/series.cpp
+    study/parts/common/cluster.cpp
+    study/parts/common/cluster.h
+    study/parts/common/cluster_list.h
```
There is no **cluster_list.hxx** in **SRC_STUDY_PART_COMMON** ?

flomnes replied:
no

guilpier-code commented on src/libs/antares/study/parts/common/cluster_list.hxx:
```diff
@@ -0,0 +1,636 @@
+#ifndef __CLUSTER_LIST_COMMON_HXX__
```
Not sure this file is useful, it seems never to be included (even in CMake files). Did I miss something ?
And the methods it defines are already defined in **cluster_list.cpp**

flomnes replied:
removed

flomnes committed:
Remove sizeGroup in ClusterList constructors

Use ClusterT::groupMax instead; with ClusterT=RenewableCluster or
ThermalCluster.

flomnes committed:
Remove "thermal" in common cluster

flomnes committed:
Remove unused function Data::ThermalCluster::FlexibilityIsValid

flomnes committed:
Move invalidateArea to common cluster class

flomnes committed:
Use generic CompareClusterName structure

flomnes committed:
Remove useless include guards

flomnes committed:
Make isVisibleOnLayer to common cluster class

flomnes committed:
Make reset method partially common

flomnes commented:
Local merge in PR #267
